File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/application/ReservationApplicationService.java
package dev.vra.poc00.application;

import dev.vra.poc00.domain.DomainFailure;
import dev.vra.poc00.domain.InventoryBalance;
import dev.vra.poc00.domain.SkuId;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ReservationApplicationService {
    private final InventoryRepository inventory;
    public ReservationApplicationService(InventoryRepository inventory) { this.inventory = inventory; }

    /** The transaction begins at this public application boundary through Spring's proxy. */
    @Transactional
    public InventoryBalance reserve(SkuId skuId, int quantity) {
        InventoryBalance.requireQuantity(quantity);
        return inventory.reserve(skuId, quantity).orElseThrow(() -> {
            if (!inventory.exists(skuId)) return new DomainFailure.UnknownSku();
            return new DomainFailure.InsufficientStock();
        });
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/application/ReservationApplicationService.kt
package dev.vra.poc00.kotlin.application

import dev.vra.poc00.kotlin.domain.DomainFailure
import dev.vra.poc00.kotlin.domain.InventoryBalance
import dev.vra.poc00.kotlin.domain.SkuId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReservationApplicationService(private val inventory: InventoryRepository) {
    /** Public proxied application boundary; only local SQL work occurs in this transaction. */
    @Transactional
    fun reserve(skuId: SkuId, quantity: Int): InventoryBalance {
        InventoryBalance.requireQuantity(quantity)
        val result = inventory.reserve(skuId, quantity)
        if (result != null) return result
        if (!inventory.exists(skuId)) throw DomainFailure.UnknownSku()
        throw DomainFailure.InsufficientStock()
    }
}
